osu.Game/Screens/Edit/HitObjectChangeHandler.cs

```diff
@@ -0,0 +1,293 @@
+        /// Redoes the last transaction from the redo stack.
+        /// Returns false if there are is nothing to redo.
+        /// </summary>
+        public bool Redo()
+        {
+            if (redoStack.Count == 0)
+                return false;
+
+            var transaction = redoStack.Pop();
+
+            if (transaction.IsChangeHandlerTransaction)
+            {
+                isRestoring = true;
+                changeHandler!.RestoreState(1);
+                isRestoring = false;
+                Logger.Log("Redo handled by old change handler");
+            }
+            else
+            {
+                applyTransaction(transaction);
+                Logger.Log("Redo handled by new change handler");
+            }
+
+            undoStack.Push(transaction);
+
+            OnStateChange?.Invoke();
+            historyChanged();
+
+            return true;
+        }
+
+        private void revertTransaction(Transaction transaction)
+        {
+            // We are navigating history so we don't want to write a new state.
+            if (changeHandler != null)
+                changeHandler.SuppressStateChange = true;
+
+            isRestoring = true;
+            editorBeatmap.BeginChange();
+
+            foreach (var change in transaction.UndoChanges.Reverse())
+                change.Revert();
+
+            foreach (var hitObject in transaction.HitObjectUpdates)
+                editorBeatmap.Update(hitObject);
+
+            editorBeatmap.EndChange();
+            isRestoring = false;
+
+            if (changeHandler != null)
+                changeHandler.SuppressStateChange = false;
+        }
+
+        private void applyTransaction(Transaction transaction)
+        {
+            // We are navigating history so we don't want to write a new state.
+            if (changeHandler != null)
+                changeHandler.SuppressStateChange = true;
+
+            isRestoring = true;
+            editorBeatmap.BeginChange();
+
+            foreach (var change in transaction.UndoChanges)
+                change.Apply();
+
+            foreach (var hitObject in transaction.HitObjectUpdates)
+                editorBeatmap.Update(hitObject);
+
+            editorBeatmap.EndChange();
+            isRestoring = false;
+
+            if (changeHandler != null)
+                changeHandler.SuppressStateChange = false;
+        }
+
+        private void historyChanged()
+        {
+            CanUndo.Value = undoStack.Count > 0;
+            CanRedo.Value = redoStack.Count > 0;
+        }
+
+        /// <summary>
+        /// Adds a change to the history but does not apply it.
+        /// </summary>
+        /// <param name="change">Change to be recorded.</param>
+        public void Record(IRevertibleChange change)
+        {
+            currentTransaction.Add(change);
+        }
+
+        public void RecordUpdate(HitObject hitObject)
+        {
+            currentTransaction.RecordUpdate(hitObject);
+        }
```

Comment:
What are these for? Why would you ever want to "record a change to history but not apply it"? What applies it in that case?